### PR TITLE
:sparkles: [core] PATCH without http-method-override header

### PR DESCRIPTION
**Pull request by bb441db**
## Description

This PR allows `PATCH` requests to be made without a `http-method-override` header

## Type of change

Check all that apply

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (a change which changes the current internal or external interface)
- [ ] This change requires a documentation update

## How Has This Been Tested?

*In case you did not include tests describe why you and how you have verified the
changes, with instructions so we can reproduce. If you have added comprehensive
tests for your changes, you may omit this section.*

**I have not created tests to verify that this PR is working, however I have modified the sample android app and I can confirm that a patch request is working with httpbin.org.
If a test case is required then I'm willing to add one (might need some help with this)**

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation, if necessary
- [ ] My changes generate no new compiler warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Inspect the bytecode viewer, including reasoning why

### CODE REVIEW

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt [hunk 1]
```diff
@@ -1,14 +1,7 @@
 package com.github.kittinunf.fuel.toolbox
 
-import com.github.kittinunf.fuel.core.Client
+import com.github.kittinunf.fuel.core.*
```
Comment (SleeplessByte) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt]: Disable * imports.

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt [hunk 2]
```diff
@@ -74,13 +67,37 @@ internal class HttpClient(
...
+     * When method is Method.PATCH, use reflection to set private field "method" on the HttpURLConnection instance
+     * For other methods, fall back to the public setter "setRequestMethod"
+     */
+    private fun HttpURLConnection.forceMethod(method: Method) {
+        when (method) {
+            Method.PATCH -> {
+                try {
+                    (this.javaClass.getDeclaredField("delegate").apply { this.isAccessible = true }.get(this) as HttpURLConnection?)?.forceMethod(method)
+                } catch (ex: NoSuchFieldException) {
+                    // ignore
```
Comment (SleeplessByte) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt]: Okay but _can_ it fail? Is this field guaranteed to be as it is on _all_ JVM implementatons of `HttpURLConnection`, if the url is yes, this comment should say that. If the answer is no, the fallback should be the header override.

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt [hunk 3]
```diff
@@ -74,13 +67,37 @@ internal class HttpClient(
...
+                }
+                (listOf(this.javaClass.superclass, this.javaClass)).forEach {
+                    try {
```
Comment (SleeplessByte) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt]: Please wrap the entire thing in `try catch` and remove the double nested try catch.

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt [hunk 4]
```diff
@@ -74,13 +67,37 @@ internal class HttpClient(
...
+    }
+
     @Throws(InterruptedException::class)
     private fun sendRequest(request: Request, connection: HttpURLConnection) {
         ensureRequestActive(request, connection)
         connection.apply {
             connectTimeout = max(request.executionOptions.timeoutInMillisecond, 0)
             readTimeout = max(request.executionOptions.timeoutReadInMillisecond, 0)
-            requestMethod = HttpClient.coerceMethod(request.method).value
+            forceMethod(request.method)
```
Comment (SleeplessByte) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt]: Should fallback to `coerce`

File: fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt [hunk 5]
```diff
@@ -110,11 +127,6 @@ internal class HttpClient(
                 Headers.collapse(HeaderName(Headers.ACCEPT_TRANSFER_ENCODING), SUPPORTED_DECODING)
             )
 
-            // The underlying HttpURLConnection does not support PATCH.
-            if (request.method == Method.PATCH) {
```
Comment (SleeplessByte) [inline fuel/src/main/kotlin/com/github/kittinunf/fuel/toolbox/HttpClient.kt]: Should be a fallback

File: sample/src/main/kotlin/com/example/fuel/MainActivity.kt [hunk 6]
```diff
@@ -4,11 +4,7 @@ import android.os.Bundle
 import android.os.Handler
 import android.support.v7.app.AppCompatActivity
 import android.util.Log
-import com.example.fuel.R.id.mainAuxText
-import com.example.fuel.R.id.mainClearButton
-import com.example.fuel.R.id.mainGoButton
-import com.example.fuel.R.id.mainGoCoroutineButton
-import com.example.fuel.R.id.mainResultText
+import com.github.kittinunf.fuel.*
```
Comment (SleeplessByte) [inline sample/src/main/kotlin/com/example/fuel/MainActivity.kt]: Disable * imports

**Comment by bb441db** [comment]
@SleeplessByte Thanks for the feedback, I have disabled the * imports and `forceMethod` now falls back to `coerceMethod` and sets the appropriate `X-HTTP-Override-method`

The multiple "try catches" are there because HttpURLConnection can have a field "delegate" of type "HttpURLConnection" and we need to call `forceMethod` on that instance aswell

I am not sure if the private field "method" is as it is on all JVM implementations.

**Comment by SleeplessByte** [comment]
@bb441db Thank you for making these changes.

These tests still fail:
```
com.github.kittinunf.fuel.toolbox.HttpClientTest > allowPatchWithBody FAILED
    java.lang.AssertionError at HttpClientTest.kt:109
com.github.kittinunf.fuel.toolbox.HttpClientTest > usesOverrideMethodForPatch FAILED
    java.lang.AssertionError at HttpClientTest.kt:38
```

**Comment by bb441db** [comment]
@SleeplessByte Tests are working for me now, I dont know why the travic-ci server is taking so long to report its status.

**Comment by SleeplessByte** [comment]
@bb441db I just need to run this against a test suite I have locally (test app) and then this is good to go. Thank you for your work!!!

**Comment by bb441db** [comment]
@SleeplessByte Since this PR is a breaking change, I think it might be better to make developers **opt-in** to this new behavior.

I have created a new branch and added this feature, let me know what you think: https://github.com/bb441db/fuel/compare/patch-hack...opt-in-force-methods

Will merge this branch if you agree with these changes.

**Comment by SleeplessByte** [comment]
Oehhh. I like it. Let me check it later tonight and merge if we can. Thanks
for this Jelle!!!

**Comment by SleeplessByte** [comment]
@bb441db yes, that will work!

**Comment by kittinunf** [comment]
I think this looks nice to me. Anything blocks this for merging?

**Comment by bb441db** [comment]
Sorry haven't had time to make these last changes, the `forceMethods` option is now passed to the client by `ExecutionOptions`

**Comment by eduardopc** [comment]
I'm very interested in this PR, I believe that'll resolve a huge problem that I'm facing right now, is there something issue to fix beyond the conflicts?

**Comment by bb441db** [comment]
@eduardopc @SleeplessByte Will merge upstream and resolve conflicts later today, this is also no longer a breaking change since the developer has to opt-**in** to this new behaviour

**Comment by kittinunf** [comment]
@bb441db does this look like a final version? I think this is a great addition to Fuel core!

**Comment by bb441db** [comment]
> @bb441db does this look like a final version? I think this is a great addition to Fuel core!

@kittinunf I think so

**Comment by kittinunf** [comment]
Let's merge this then? @SleeplessByte do you have any objection, sir?

**Comment by SleeplessByte** [comment]
Nope. Let's do this!

**Pull request merged**